pallets/pooled-staking/src/candidate.rs

```diff
@@ -0,0 +1,182 @@
+// Copyright (C) Moondance Labs Ltd.
+// This file is part of Tanssi.
+
+// Tanssi is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Tanssi is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Tanssi.  If not, see <http://www.gnu.org/licenses/>
+
+use {
+    crate::{
+        pools::{self, Pool},
+        traits::{ErrAdd, ErrSub},
+        Candidate, Config, Error, Event, Pallet, Pools, PoolsKey, SortedEligibleCandidates, Stake,
+    },
+    core::{cmp::Ordering, marker::PhantomData},
+    parity_scale_codec::{Decode, Encode},
+    scale_info::TypeInfo,
+    sp_core::{Get, RuntimeDebug},
+    sp_runtime::traits::Zero,
+};
+
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
+
+/// Eligible candidate with its stake.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(RuntimeDebug, PartialEq, Eq, Encode, Decode, Clone, TypeInfo)]
+pub struct EligibleCandidate<C, S> {
+    candidate: C,
+    stake: S,
+}
+
+impl<C: Ord, S: Ord> Ord for EligibleCandidate<C, S> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.stake
+            .cmp(&other.stake)
+            .reverse()
+            .then(self.candidate.cmp(&self.candidate))
```

Comment:
This is comparing self to self, should be other.

Also this can benefit from lazy evaluation, because the stake amount is unlikely to be equal (using `then(` will always compare it, `and_then` only if previous condition is equal):

```suggestion
            .and_then(|| self.candidate.cmp(&other.candidate))
```